### PR TITLE
Add spec text for groupBy

### LINKED ISSUE #15: groupBy spec text
In the slides for this meeting, there is a request for a new proposal, groupBy, to be moved to stage 2 as well. However I am unable to find spec text for it. Is it published anywhere?

### PR DESCRIPTION
**Pull request by jridgewell**
Closes #15

### CODE REVIEW

File: spec.emu [hunk 1]
```diff
@@ -56,4 +56,44 @@ location: https://tc39.es/proposal-array-filtering/
...
+      <p>`groupBy` does not directly mutate the object on which it is called but the object may be mutated by the calls to _callbackfn_.</p>
+      <p>The range of elements processed by `groupBy` is set before the first call to _callbackfn_. Elements which are appended to the array after the call to `groupBy` begins will not be visited by _callbackfn_. If existing elements of the array are changed their value as passed to _callbackfn_ will be the value at the time `groupBy` visits them; elements that are deleted after the call to `groupBy` begins and before being visited are not visited.</p>
+    </emu-note>
+    <p>When the `groupBy` method is called with one or two arguments, the following steps are taken:</p>
+    <emu-alg>
+      1. Let _O_ be ? ToObject(*this* value).
+      1. Let _len_ be ? LengthOfArrayLike(_O_).
+      1. If IsCallable(_callbackfn_) is *false*, throw a *TypeError* exception.
+      1. Let _k_ be 0.
+      1. Let _obj_ be ! OrdinaryObjectCreate(%Object.prototype%).
```
Comment (zloirock) [inline spec.emu:76]: Maybe `null` prototype could be better?

Comment (jridgewell) [reply spec.emu:76]: Quite possibly, but this would need to be discussed in plenary. I'll add it to the slides, thanks!

Comment (ljharb) [reply spec.emu:76]: The spec typically doesn't create null objects; module namespace objects and Object.create(null) are outliers. As long as users use Object.keys/entries/values/assign/object spread with these objects, or direct property access of expected groups (and thus, present as own properties) - all of which are what I'd expect the entirety of use cases to be - it wouldn't make a difference if it had a null proto or not.

Comment (zloirock) [reply spec.emu:76]: In some cases it's used in the spec - for example, `RegExp` named groups - and it's a similar case.

Users will use it as
```js
const { foo = [], bar = [], baz = [] } = array.groupBy(fn);
```

It makes no sense to force users to check `Object.hasOwn` each time for preventing
```js
Object.prototype.foo = ['oops'];
```
->
```js
const result = array.groupBy(fn);
const foo = Object.hasOwn(result, 'foo') ? result.foo : [];
const bar = Object.hasOwn(result, 'bar') ? result.bar : [];
const baz = Object.hasOwn(result, 'baz') ? result.baz : [];
```

Comment (ljharb) [reply spec.emu:76]: Why would they need to? They’ll know which properties they’ve just grouped by.

Comment (zloirock) [reply spec.emu:76]: Some groups can haven't any elements -> missed on the result.

File: spec.emu [hunk 2]
```diff
@@ -56,4 +56,44 @@ location: https://tc39.es/proposal-array-filtering/
...
+      1. Assert: _obj_ is an extensible ordinary object with no own properties.
+      1. Repeat, while _k_ &lt; _len_
+        1. Let _Pk_ be ! ToString(_k_).
+        1. Let _kPresent_ be ? HasProperty(_O_, _Pk_).
```
Comment (zloirock) [inline spec.emu:80]: It's a controversial moment. Modern methods haven't this check.

Comment (jridgewell) [reply spec.emu:80]: Can you point to an example? `flatMap` is the most recent addition that I can think of, and it has one in https://tc39.es/ecma262/#sec-flattenintoarray at 6.b and 6.c.

Comment (zloirock) [reply spec.emu:80]: It was decided for ES6+ methods. `.find`, `.findIndex`, `.includes`, iterators, etc. `FlattenIntoArray` is another case and I think that this check was added to the spec [mistakenly](https://github.com/tc39/proposal-flatMap/issues/29).

Comment (jridgewell) [reply spec.emu:80]: I added this to the slides to discuss.

**Pull request merged**